packages/runtime/src/runtimes/react/controls.tsx

```diff
@@ -1,164 +1,41 @@
 import { useRef } from 'react'
 
-import * as ReactPage from '../../state/react-page'
-
-import {
-  CheckboxDefinition,
-  NumberDefinition,
-  RichTextV2Definition,
-  RichTextV2Control,
-  ColorDefinition,
-  ComboboxDefinition,
-  IconRadioGroupDefinition,
-  ImageDefinition,
-  LinkDefinition,
-  ListDefinition,
-  RichTextV1Definition,
-  RichTextV1Control,
-  SelectDefinition,
-  ShapeDefinition,
-  SlotDefinition,
-  SlotControl,
-  StyleDefinition,
-  StyleControl,
-  StyleV2Definition,
-  TextAreaDefinition,
-  TextInputDefinition,
-  unstable_TypographyDefinition,
-} from '../../controls'
+import { ControlDefinition } from '@makeswift/controls'
 
-import { isLegacyDescriptor } from '../../prop-controllers/descriptors'
+import { partitionRecord } from '../../utils/partition'
 
-import { useFormattedStyle } from './controls/style'
-import { ControlValue } from './controls/control'
-import { useSlot } from './controls/slot'
-import { useRichText } from './controls/rich-text/rich-text'
-import { useRichTextV2 } from './controls/rich-text-v2'
+import * as ReactPage from '../../state/react-page'
+import { isLegacyDescriptor, LegacyDescriptor } from '../../prop-controllers/descriptors'
 
 import { useStore } from './hooks/use-store'
-import { useDocumentKey } from './hooks/use-document-key'
-import { useSelector } from './hooks/use-selector'
 
-import { RenderHook } from './components'
 import { resolveLegacyDescriptorProp } from './legacy-controls'
+import { useResolvedProps } from './hooks/use-resolved-props'
 
 type PropsValueProps = {
   element: ReactPage.ElementData
   children(props: Record<string, unknown>): JSX.Element
 }
 
-export function PropsValue({ element, children }: PropsValueProps): JSX.Element {
+function useControlDefs(
+  elementType: string,
+): readonly [Record<string, LegacyDescriptor>, Record<string, ControlDefinition>] {
   const store = useStore()
-  const propControllerDescriptorsRef = useRef(
-    ReactPage.getComponentPropControllerDescriptors(store.getState(), element.type) ?? {},
-  )
-  const props = element.props as Record<string, any>
-  const documentKey = useDocumentKey()
+  const all = ReactPage.getComponentPropControllerDescriptors(store.getState(), elementType) ?? {}
+  return useRef(partitionRecord(all, isLegacyDescriptor)).current
+}
 
-  const propControllers = useSelector(state => {
-    if (documentKey == null) return null
+export function ResolveProps({ element, children: renderComponent }: PropsValueProps): JSX.Element {
+  const [legacyDescriptors, definitions] = useControlDefs(element.type)
 
-    return ReactPage.getPropControllers(state, documentKey, element.key)
-  })
+  const resolvedProps = useResolvedProps(definitions, element.props, element.key)
 
-  return Object.entries(propControllerDescriptorsRef.current).reduceRight(
+  const renderFn = Object.entries(legacyDescriptors).reduceRight(
```

Comment:
Properties coming from legacy descriptors aka prop controllers are still resolved as before, outside of the unified controls framework.